src/librbd/cache/pwl/Request.h

```diff
@@ -170,14 +170,18 @@ class C_WriteRequest : public C_BlockIORequest<T> {
     return "C_WriteRequest";
   }
 
+  virtual void setup_pmem_buffers(
```

Comment:
Nit: why do we need this `pmem` method in a generic request?

Reply:
Here it's just renamed. This function handles the RWL case and hasn't been moved out. `Request` class could be split for SSD and RWL but that may result in some code duplication. For example, `C_CompAndWriteRequest` inherits from `C_WriteRequest` and if `setup_buffer_resources` (or `setup_pmem_resources`) function is moved to `C_SSDWriteRequest` (and corresponding C_RWLWriteRequest class), then `C_CompAndWriteRequest` won't have access to it and we may have to create `C_SSDCompAndWriteRequest` (and an RWL equivalent) and copy the code from `setup_buffer_resources` -- would that be ok?

Reply:
The base `C_CompAndWriteRequest` class could be virtual and the `C_RWLWriteRequest` version could invoke the base class version and then prepare the PMEM buffers. It's not like the current `ReplicatedWriteLog<I>::setup_resources` is doing much besides bouncing the function call back.

Reply:
Not sure I understand. Currently Request class is not split for RWL and SSD. But if we were to split, it'd look like

             C_WriteRequest {virtual void setup_pmem_buffers() {} //overridden in derived classes)
                      | 
                     / \ 
    C_SSDWriteRequest    C_RWLWriteRequest 

But if `C_RWLWriteRequest` instead inherits from `C_CompAndWriteRequest` and then implements `setup_pmem_buffers`, how can a `C_CompAndWriteRequest` object access it? Wouldn't we still end up at the same problem?

Reply:
I think you meant inheritance like below?

             C_WriteRequest 
                     |
             C_CompAndWriteRequest
                     |
            C_RWLWriteRequest
                     |
           C_RWLCompAndWriteRequest

I think the above will work.

Reply:
Well, the goal I think would be to split `C_WriteRequest` to avoid having pmem-specific things in the base class. One possible way to handle that would be to move the pmem-specific logic to its own class and then the hierarchy becomes:

```
C_WriteRequest<Payload>
          |
C_CompAndWriteRequest<Payload>
```

Where RWL would have some pmem-related payload available which you could either provide to the RWL class to invoke whatever specialized methods it needs, or if it's generic enough between SSD and RWL, `setup_resources` could invoke the payload setup method directly.

Reply:
This has been addressed. But trying to see if there's a clean and generic way to remove references of pmem from base classes